File: ios/flutter_card_scanner_plus/Sources/flutter_card_scanner_plus/NfcSession.swift
import CoreNFC
import Flutter
import Foundation

/// Thin bridge to Core NFC: open a reader session, exchange APDUs, close.
///
/// No EMV logic lives here. The Dart side decides what to send and how to
/// read the reply, which keeps that logic testable without a device.
@available(iOS 13.0, *)
final class NfcSession: NSObject, NFCTagReaderSessionDelegate {
  private var session: NFCTagReaderSession?
  private var tag: NFCISO7816Tag?

  /// Completion of the pending `connect` call. Cleared once answered so it
  /// can never be called twice.
  private var connectResult: FlutterResult?
  private let lock = NSLock()

  static var isAvailable: Bool { NFCTagReaderSession.readingAvailable }

  func connect(prompt: String, result: @escaping FlutterResult) {
    guard NFCTagReaderSession.readingAvailable else {
      result(
        FlutterError(
          code: "nfcUnavailable",
          message: "NFC tag reading is not available on this device",
          details: nil))
      return
    }
    close(message: nil, errorMessage: nil)

    guard
      let session = NFCTagReaderSession(pollingOption: [.iso14443], delegate: self, queue: nil)
    else {
      result(
        FlutterError(
          code: "nfcSessionError", message: "Could not start a reader session", details: nil))
      return
    }
    lock.lock()
    connectResult = result
    lock.unlock()

    session.alertMessage = prompt
    self.session = session
    session.begin()
  }

  func transceive(command: Data, result: @escaping FlutterResult) {
    guard let tag else {
      result(FlutterError(code: "nfcTagLost", message: "No card is connected", details: nil))
      return
    }
    guard let apdu = NFCISO7816APDU(data: command) else {
      result(FlutterError(code: "invalidArgument", message: "Malformed APDU", details: nil))
      return
    }
    tag.sendCommand(apdu: apdu) { payload, sw1, sw2, error in
      if let error {
        result(
          FlutterError(
            code: "nfcTagLost", message: error.localizedDescription, details: nil))
        return
      }
      var reply = payload
      reply.append(sw1)
      reply.append(sw2)
      result(FlutterStandardTypedData(bytes: reply))
    }
  }

  func close(message: String?, errorMessage: String?) {
    answerConnect(
      FlutterError(code: "nfcCancelled", message: "Session closed", details: nil))
    if let session {
      if let errorMessage {
        session.invalidate(errorMessage: errorMessage)
      } else {
        if let message { session.alertMessage = message }
        session.invalidate()
      }
    }
    session = nil
    tag = nil
  }

  /// Calls the pending connect completion exactly once.
  private func answerConnect(_ value: Any?) {
    lock.lock()
    let pending = connectResult
    connectResult = nil
    lock.unlock()
    pending?(value)
  }

  // MARK: - NFCTagReaderSessionDelegate

  func tagReaderSessionDidBecomeActive(_ session: NFCTagReaderSession) {}

  func tagReaderSession(_ session: NFCTagReaderSession, didInvalidateWithError error: Error) {
    let code: String
    if let readerError = error as? NFCReaderError,
      readerError.code == .readerSessionInvalidationErrorUserCanceled
        || readerError.code == .readerSessionInvalidationErrorSessionTimeout
    {
      code = "nfcCancelled"
    } else {
      code = "nfcSessionError"
    }
    answerConnect(
      FlutterError(code: code, message: error.localizedDescription, details: nil))
    self.session = nil
    tag = nil
  }

  func tagReaderSession(_ session: NFCTagReaderSession, didDetect tags: [NFCTag]) {
    guard let first = tags.first, case let .iso7816(isoTag) = first else {
      session.invalidate(errorMessage: "This is not a payment card")
      answerConnect(
        FlutterError(
          code: "nfcSessionError", message: "Tag is not ISO 7816", details: nil))
      return
    }

    session.connect(to: first) { [weak self] error in
      guard let self else { return }
      if let error {
        session.invalidate(errorMessage: "Could not connect to the card")
        self.answerConnect(
          FlutterError(
            code: "nfcTagLost", message: error.localizedDescription, details: nil))
        return
      }
      self.tag = isoTag
      self.answerConnect(nil)
    }
  }
}

File: android/src/main/kotlin/dev/apissystems/flutter_card_scanner_plus/NfcSession.kt
package dev.apissystems.flutter_card_scanner_plus

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.IsoDep
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel

/**
 * Thin bridge to `android.nfc`: open a reader session, exchange APDUs, close.
 *
 * No EMV logic lives here. The Dart side decides what to send and how to read
 * the reply, which keeps that logic testable without a device.
 *
 * The plugin deliberately does not declare `android.permission.NFC` in its
 * manifest, so apps that never read cards do not inherit it. An app that wants
 * NFC adds the permission itself; without it this reports `nfcUnavailable`.
 */
class NfcSession(private val context: Context) {

    private val main = Handler(Looper.getMainLooper())
    private var activity: Activity? = null
    private var isoDep: IsoDep? = null
    private var pending: MethodChannel.Result? = null

    fun attach(activity: Activity?) {
        this.activity = activity
    }

    fun isAvailable(): Boolean {
        if (!hasPermission()) return false
        val adapter = NfcAdapter.getDefaultAdapter(context) ?: return false
        return adapter.isEnabled
    }

    fun connect(result: MethodChannel.Result) {
        val activity = activity
        if (activity == null) {
            result.error("nfcSessionError", "Plugin is not attached to an Activity", null)
            return
        }
        if (!hasPermission()) {
            result.error(
                "nfcUnavailable",
                "Add <uses-permission android:name=\"android.permission.NFC\" /> to the app manifest",
                null,
            )
            return
        }
        val adapter = NfcAdapter.getDefaultAdapter(context)
        if (adapter == null || !adapter.isEnabled) {
            result.error("nfcUnavailable", "NFC is off or not supported", null)
            return
        }

        close(null, null)
        pending = result
        try {
            adapter.enableReaderMode(
                activity,
                ::onTag,
                NfcAdapter.FLAG_READER_NFC_A or
                    NfcAdapter.FLAG_READER_NFC_B or
                    NfcAdapter.FLAG_READER_SKIP_NDEF_CHECK,
                null,
            )
        } catch (e: SecurityException) {
            pending = null
            result.error("nfcUnavailable", e.message, null)
        }
    }

    private fun onTag(tag: Tag) {
        val dep = IsoDep.get(tag)
        if (dep == null) {
            answer { it.error("nfcSessionError", "Tag is not ISO 7816", null) }
            return
        }
        try {
            dep.timeout = TIMEOUT_MS
            dep.connect()
            isoDep = dep
            answer { it.success(null) }
        } catch (e: Exception) {
            answer { it.error("nfcTagLost", e.message, null) }
        }
    }

    fun transceive(command: ByteArray, result: MethodChannel.Result) {
        val dep = isoDep
        if (dep == null || !dep.isConnected) {
            result.error("nfcTagLost", "No card is connected", null)
            return
        }
        // IsoDep.transceive blocks, so it must not run on the main thread.
        Thread {
            try {
                val reply = dep.transceive(command)
                main.post { result.success(reply) }
            } catch (e: Exception) {
                main.post { result.error("nfcTagLost", e.message, null) }
            }
        }.start()
    }

    fun close(message: String?, errorMessage: String?) {
        answer { it.error("nfcCancelled", "Session closed", null) }
        try {
            isoDep?.close()
        } catch (_: Exception) {
            // Already gone.
        }
        isoDep = null
        val activity = activity ?: return
        if (!hasPermission()) return
        try {
            NfcAdapter.getDefaultAdapter(context)?.disableReaderMode(activity)
        } catch (_: Exception) {
            // Nothing to disable.
        }
    }

    /** Calls the pending connect completion exactly once, on the main thread. */
    private fun answer(block: (MethodChannel.Result) -> Unit) {
        val result = pending ?: return
        pending = null
        main.post { block(result) }
    }

    private fun hasPermission(): Boolean =
        context.checkPermission(
            Manifest.permission.NFC,
            android.os.Process.myPid(),
            android.os.Process.myUid(),
        ) == PackageManager.PERMISSION_GRANTED

    private companion object {
        const val TIMEOUT_MS = 5_000
    }
}
